README.md
# ls-proxy [![Build Badge]](https://gitlab.com/MysteryBlokHed/ls-proxy/-/pipelines) [![NPM Badge]](https://www.npmjs.com/package/ls-proxy) [![License Badge]](#license)

Wrapper around localStorage (**[and other stores](#custom-stores)**) to easily store JSON objects.

Supports:

- Storing any type that can be serialized
- Runtime valdiation (eg. type checking)
- Deeply nested objects
- Defining custom stores other than localStorage

## Why to use?

### For localStorage

If you want to store client-side data for your website, the way to do it is with localStorage.
However, there is at least one siginificant downside: you can only store strings in localStorage keys.

The best way to get around this is by storing a stringified JSON object in a key,
but doing this manually or having to call a function that does it for you any time you change an object would be annoying.

This library solves these problems using JS proxies.
It also has great IDE support thanks to it being written in TypeScript.
You can also use it with vanilla JS with the Webpacked file (`ls-proxy.user.js` or `ls-proxy.min.user.js`),
which is useful to test it in the browser, when not using a JS bundler, or while writing UserScripts.

Here's all it takes to store a stringifed JSON object in localStorage and automatically change it:

```typescript
import { storeObject } from 'ls-proxy'

const someInfo = storeObject(
  // The localStorage key to save data under
  'someInfo',
  // The object to store
  {
    aString: 'Hello, World!',
    aNumber: 123,
    aBoolean: true,
    aList: [1, '2', 3],
  },
)

someInfo.aNumber = 42 // Updates localStorage
console.log(someInfo.aList) // Reads from localStorage
```

The method above stores an entire serialized object in localStorage,
meaning the entire object has to be stringified/parsed whenever a single key
is affected. The method below stores each key individually instead:

```typescript
import { storeSeparate } from 'ls-proxy'

const someInfo = storeObject(
  // The object to store
  {
    aString: 'Hello, World!',
    aNumber: 123,
    aBoolean: true,
    aList: [1, '2', 3],
  },
  // Optional; prefixes the stored keys with this ID
  { id: 'someInfo' },
)

someInfo.aNumber = 42 // Updates localStorage
console.log(someInfo.aList) // Reads from localStorage
```

### For other stores

If you have any value stored in a store that requires some extra action to trigger an update,
then using ls-proxy to create a custom store is a good option.
One example of this is React (and many other frontend UI's) state, which has to be updated by calling
a `setState` function instead of just mutating the value returned.
When using `ls-proxy`, state will be updated even if deeply nested objects' keys are modified.

You can override the `get` and `set` methods to automatically set/retrieve values from your other store:

```typescript
import { storeSeparate } from 'ls-proxy'

const myObj = storeSeparate({ foo: 'bar' },
{
  get(key) {
    // Logic to get a key here
  }

  set(key, value) {
    // Logic to set a key here
  }

  // Important!!
  // If your store only accepts strings, then don't override this method
  // If it accepts other values such as objects, then override it as shown
  // This makes sure that the proxy created by ls-proxy isn't accidentally stored,
  // which could cause a lot of very painful-to-find bugs
  stringify(value) {
    if (typeof value == 'object') return JSON.parse(JSON.stringify(value))
    return value
  }
})
```

#### Making a resuable function

To make a function to construct these objects, it's a good idea to still let the user pass in config options.
You can copy the signature from `storeObject` or `storeSeparate` as long as it's credited as described
by the project's licenses.

Here's an example:

```typescript
import { storeSeparate, StoreSeparateOptions } from 'ls-proxy'

// The options for your store function
// This should extend the original function's options, omitting those that your function overrides
type Options<O extends Record<string, any>> = Omit<
  StoreSeparateOptions<O>,
  'get' | 'set'
>

function storeCustom<O extends Record<string, any>>(
  defaults: O,
  configuration: Options<O>,
): O {
  return storeSeparate(defaults, {
    ...configuration,
    get(key) {
      // Logic to get a key here
    },

    set(key, value) {
      // Logic to set a key here
    },
  })
}
```

## Documentation

Documentation for the main branch is hosted at <https://ls-proxy.adamts.me>.
Documentation can be built from a cloned repository by running `yarn doc`.

Some examples are located in [`examples`](https://gitlab.com/MysteryBlokHed/ls-proxy/-/tree/main/examples),
but many others are available in the JSDoc for their respective functions.

## storeObject vs storeSeparate

`storeSeparate` will generally be faster than `storeObject`
since there's less data being serialized/deserialized with each get/set,
but there are still reasons to use `storeObject`.
For example, if you want to use `validate` and `modify` (see documentation for config options),
and you need the entire object for context.
This can be the case if you need another key's value to validate the object,
or if you want to modify multiple keys with a single set/get.

## Custom stores

Because of the customization offered in the options of `storeObject` and `storeSeparate`,
it's possible to define custom stores other than localStorage.

These can be made by overriding the `get` and `set` methods via the config options.
There are some custom stores already built in in the `factories` subfolder:

- React State for functional components (`useStateProxy` in `ls-proxy/factories/react`)
  - Allows state to be used in a way more similar to that of class components
    in the context of a functional component
- SolidJS Signals (`createSignalProxy` in `ls-proxy/factories/solid`)
  - Achieves a similar effect as the React factory

See [For other stores](#for-other-stores) for a brief explanation about how to set one up.

## Use

### In a Node project

To use in a Node project, add ls-proxy as a dependency.

```sh
# npm
npm install ls-proxy

# yarn
yarn add ls-proxy
```

You can then import and use functions:

```javascript
import { storeObject } from 'ls-proxy'

const myObj = storeObject('myObj', {
  name: 'John',
  age: 21,
})

myObj.name = 'Joe'
myObj.age++
```

### In a normal UserScript

In a UserScript that isn't built with some build tool, you can `@require` the library:

```javascript
// @require     https://gitlab.com/MysteryBlokHed/ls-proxy/-/raw/main/ls-proxy.user.js
```

<!-- Make sure that this is true for your project -->

You can replace `main` with a specific release tag like `v0.1.0` to require a specific version:

```javascript
// @require     https://gitlab.com/MysteryBlokHed/ls-proxy/-/raw/v0.1.0/ls-proxy.user.js
```

Each release tag also has a minified version of the script available,
which can be used by changing the file extension to `.min.user.js`:

```javascript
// @require     https://gitlab.com/MysteryBlokHed/ls-proxy/-/raw/v0.1.0/ls-proxy.min.user.js
```

Functions are available on the global `LSProxy` object:

```javascript
const { storeObject } = LSProxy

const myObj = storeObject('myObj', {
  name: 'John',
  age: 21,
})

myObj.name = 'Joe'
myObj.age++
```

#### Type declarations

The types included with the npm package still work when the library is `@require`'d.
Just add the types as a dev dependency for a Node project or install them globally.
With the package installed, include the following reference line somewhere in your TypeScript source file:

```typescript
/// <reference types="ls-proxy" />
```

## Building

### Setup

Building this project requires Node.js and Yarn.
To install dependencies, run:

```sh
yarn install
```

### Build

To build the project, run:

```sh
yarn build
```

To automatically build when a source file is modified, run:

```sh
yarn dev
```

Built JS files and type declarations will be placed in the `lib/` directory,
and the UserScript will be placed in the root. The `package.json` file is configured
to publish files in the `lib/` directory to npm.

### Test

To test the project, run:

```sh
yarn test
```

This project uses Jest for tests.

## License

This project is licensed under either of

- Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or
  <http://www.apache.org/licenses/LICENSE-2.0>)
- MIT license ([LICENSE-MIT](LICENSE-MIT) or
  <http://opensource.org/licenses/MIT>)

at your option.

This project was created from [a template](https://gitlab.com/MysteryBlokHed/webpack-ts-userscript-library)
licensed under the MIT license
([LICENSE](https://gitlab.com/MysteryBlokHed/webpack-ts-userscript-library/-/blob/main/LICENSE)
or <http://opensource.org/licenses/MIT>).

[build badge]: https://img.shields.io/gitlab/pipeline-status/MysteryBlokHed/ls-proxy
[npm badge]: https://img.shields.io/npm/v/ls-proxy
[license badge]: https://img.shields.io/badge/license-MIT%20OR%20Apache--2.0-green


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## Changed

- Changed test imports to use `src/` instead of `lib/`

## Fixed

- Fixed incorrect function names in docs for `useStateProxy` and `createSignalProxy`

## 0.7.0

## Added

- Added `mutateProxiedObject` option to `storeObject` and `storeSeparate`
- Added a factory for SolidJS (`createSignalProxy` in `ls-proxy/factories/solid`)

## Changed

- Removed factories re-export from `index.ts`
- Renamed React factory from `storeStateful` to `useStateProxy`

## 0.6.1

### Fixed

- Configured `package.json` so that factories can be imported as shown in the docs
- Fixed comment in documentation for `storeStateful`

## 0.6.0

### Added

- Added factory functions for custom stores as well as documentation about them
  (the ones that exist and how to make them)

### Fixed

- Updated dependency on terser to non-vulnerable version
- Fixed incorrect documentation for base64 encoding example

## 0.5.1

### Fixed

- Updated some dependencies to non-vulnerable versions

## 0.5.0

### Changed

- Completely changed the `storeSeparate` function
  to automatically serialize/deserialize keys in localStorage like `storeObject`
  - Values are automatically serialized/deserialized
  - Supports `validate`, `modify`, and other functions from `storeObject`
  - Supports deeply nested objects

## 0.4.0

### Added

- Added option to define get/set functions for storeObject
- Added JSDoc for functions' generic types

## 0.3.1

### Fixed

- Fixed arrays not updating localStorage values the way objects do

## 0.3.0

### Added

- Added support for deeply nested objects
- Added action parameter to storeObject's modify and validate functions to indicate whether an object
  is being set to or retrieved from localStorage

## 0.2.1

### Changed

- Changed license from MIT to dual-license MIT OR Apache-2.0

### Fixed

- Fixed outdated JSDoc for validate function

## 0.2.0

### Added

- Option to store multiple objects under the same localStorage key (see `partial` parameter for `StoreObjectConfig`
  or `different-keys-same-object.ts` in examples)
- Testing with Jest
- Default functions for common validation use cases (key validation and basic type checking).
  See `Validations` namespace

### Changed

- Move modification ability of `validate` to new function for `storeObject`
- Removed some unecessary generics from functions

## 0.1.0

### Added

- Function to automatically change a serialized object in localStorage when an object is modified (`storeObject`)
- Function to automatically change keys in localStorage when an object is modified (`storeSeparate`)


ls-proxy.user.js
// ==UserScript==
// @name        ls-proxy
// @descripton  Wrapper around localStorage (and other stores) to easily store JSON objects
// @version     0.7.0
// @author      Adam Thompson-Sharpe
// @license     MIT OR Apache-2.0
// @homepageURL https://gitlab.com/MysteryBlokHed/ls-proxy
// ==/UserScript==

/******/ (() => { // webpackBootstrap
/******/ 	"use strict";
/******/ 	var __webpack_modules__ = ({

/***/ "./lib/validations.js":
/*!****************************!*\
  !*** ./lib/validations.js ***!
  \****************************/
/***/ ((__unused_webpack_module, exports) => {


Object.defineProperty(exports, "__esModule", ({ value: true }));
/** Validations meant to be used with `storeObject`'s validate function */
var Validations;
(function (Validations) {
    /**
     * Validate that only expected keys are present on an object
     *
     * @param value The unknown value to validate types of
     * @param requiredKeys The **only** keys that should be present
     * @template O The stored object
     *
     * @example
     * ```typescript
     * import { storeObject, Validations } from 'ls-proxy'
     *
     * const myObj = storeObject(
     *   'myObj',
     *   { foo: 'bar' },
     *   { validate: value => Validations.keys(value, ['foo']) },
     * )
     *
     * myObj.foo = 'abc' // no error
     * myObj.bar = 'xyz' // error
     * ```
     */
    Validations.keys = (value, requiredKeys) => Object.keys(value).every(key => requiredKeys.includes(key)) &&
        requiredKeys.every(key => key in value);
    /**
     * Validate that the types passed for an object are expected
     *
     * @param value The unknown value to validate types of
     * @param typesMap A map of expected keys for an object to expected types, checked like `typeof value[key] === typesMap[key]`
     * @template O The stored object
     * @example
     * ```typescript
     * import { storeObject, Validations } from 'ls-proxy'
     *
     * const typesMap = {
     *   onlyString: 'string',
     *   onlyNumber: 'number',
     * }
     *
     * const runtimeCheckedTypes = storeObject(
     *   'runtimeCheckedTypes',
     *   {
     *     onlyString: 'abc',
     *     onlyNumber: 123,
     *   },
     *   { validate: value => Validations.types(value, typesMap) },
     * )
     *
     * runtimeCheckedTypes.onlyString = 'xyz' // Succeeds
     * runtimeCheckedTypes.onlyNumber = 'abc' // Fails
     * ```
     */
    Validations.types = (value, typesMap) => Object.entries(value).every(([key, value]) => typeof value === typesMap[key]);
})(Validations || (Validations = {}));
exports["default"] = Validations;


/***/ })

/******/ 	});
/************************************************************************/
/******/ 	// The module cache
/******/ 	var __webpack_module_cache__ = {};
/******/ 	
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/ 		// Check if module is in cache
/******/ 		var cachedModule = __webpack_module_cache__[moduleId];
/******/ 		if (cachedModule !== undefined) {
/******/ 			return cachedModule.exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = __webpack_module_cache__[moduleId] = {
/******/ 			// no module.id needed
/******/ 			// no module.loaded needed
/******/ 			exports: {}
/******/ 		};
/******/ 	
/******/ 		// Execute the module function
/******/ 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
/******/ 	
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// This entry need to be wrapped in an IIFE because it need to be isolated against other modules in the chunk.
(() => {
var exports = __webpack_exports__;
/*!**********************!*\
  !*** ./lib/index.js ***!
  \**********************/

Object.defineProperty(exports, "__esModule", ({ value: true }));
exports.storeSeparate = exports.storeObject = exports.Validations = void 0;
var validations_1 = __webpack_require__(/*! ./validations */ "./lib/validations.js");
Object.defineProperty(exports, "Validations", ({ enumerable: true, get: function () { return validations_1.default; } }));
/**
 * Fill in default values for CommonConfig
 */
const commonDefaults = ({ checkGets, checkDefaults, mutateProxiedObject, set, get, parse, stringify, }) => ({
    checkGets: checkGets !== null && checkGets !== void 0 ? checkGets : true,
    checkDefaults: checkDefaults !== null && checkDefaults !== void 0 ? checkDefaults : true,
    mutateProxiedObject: mutateProxiedObject !== null && mutateProxiedObject !== void 0 ? mutateProxiedObject : true,
    set: set !== null && set !== void 0 ? set : ((key, value) => (localStorage[key] = value)),
    get: get !== null && get !== void 0 ? get : (value => { var _a; return (_a = localStorage[value]) !== null && _a !== void 0 ? _a : null; }),
    parse: parse !== null && parse !== void 0 ? parse : JSON.parse,
    stringify: stringify !== null && stringify !== void 0 ? stringify : JSON.stringify,
});
/**
 * Fill in default values for StoreObjectConfig
 * @template O The stored object
 */
const defaultStoreObjectConfig = ({ checkGets, partial, set, get, validate, modify, parse, stringify, }) => (Object.assign({ partial: partial !== null && partial !== void 0 ? partial : false, validate: validate !== null && validate !== void 0 ? validate : (() => true), modify: modify !== null && modify !== void 0 ? modify : (value => value) }, commonDefaults({ checkGets, set, get, parse, stringify })));
const shouldObjectProxy = (object) => 
// Check that the target isn't falsey (primarily in case it's null, since typeof null === 'object')
object &&
    // Check type
    typeof object === 'object' &&
    // 'object' type includes some unwanted types, so check constructor
    [Object, Array].includes(object.constructor);
/**
 * Proxy handler for deeply nested objects on the main object
 * @template P The parent object
 * @template N The child of the parent
 */
const nestedProxyHandler = (parent, parentKey, nested, parentSet) => {
    return new Proxy(nested, {
        set(target, key, value) {
            const setResult = Reflect.set(target, key, value);
            // Trigger set trap of original object, updating localStorage
            parentSet(parent, parentKey, target, parent);
            return setResult;
        },
        get(target, key) {
            if (shouldObjectProxy(target[key])) {
                // Return a Proxy to the object to catch sets
                return nestedProxyHandler(target, key, Reflect.get(target, key), this.set);
            }
            return Reflect.get(target, key);
        },
    });
};
const validateResult = (valid, defaultError) => {
    // Throw error on failure
    if (typeof valid === 'boolean') {
        // Return is bool
        if (!valid)
            return [valid, defaultError];
        return [valid];
    }
    else {
        // Return is array
        if (!valid[0]) {
            if (valid.length === 2)
                return [valid[0], valid[1]];
            else
                return [valid[0], defaultError];
        }
        return [valid[0]];
    }
};
/**
 * Validate and modify an object
 *
 * @param validate Function to validate the object
 * @param modify Function to modify the object
 * @param object The object to modify
 * @param action Whether the object is being get or set
 * @param lsKey The key in localStorage
 * @template O The stored object
 * @returns The object if valid
 */
const validOrThrow = (validate, modify, object, action, lsKey) => {
    const error = new TypeError(action === 'get'
        ? `Validation failed while parsing ${lsKey} from localStorage`
        : `Validation failed while setting to ${lsKey} in localStorage`);
    const valid = validateResult(validate(object, action), error);
    if (!valid[0])
        throw valid[1];
    return modify(object, action);
};
/**
 * Store a stringified JSON object in localStorage.
 * This method can use any type that can be serialized.
 * The object stored in localStorage is **not** checked for validity by default,
 * but you can pass an argument with your own function to do so
 *
 * @param lsKey The localStorage key to store the stringified object in
 * @param defaults The default values if the object is not stored
 * @param configuration Config options
 * @template O The stored object
 *
 * @example
 * ```typescript
 * // No validation
 * import { storeObject } from 'ls-proxy'
 *
 * // Stored serialized in localStorage under the key `myObject`
 * // Note that types other than string can be used
 * const myPerson = storeObject('myObject', {
 *   name: 'John',
 *   age: 21,
 *   interests: ['programming'],
 * })
 *
 * myPerson.name = 'Joe' // Updates localStorage
 * console.log(myPerson.name) // Checks localStorage if checkGets is true
 * ```
 *
 * @example
 * ```typescript
 * // Validating that the expected keys exist and are the correct type
 * import { storeObject, validateKeys } from 'ls-proxy'
 *
 * const myObj = storeObject(
 *   'myObj',
 *   {
 *     someString: 'string',
 *     someNumber: 42,
 *   },
 *   {
 *     validate(value) {
 *       if (!validateKeys(value, ['someString', 'someNumber'])) return false
 *       if (typeof value.someString !== 'string') return false
 *       if (typeof value.someNumber !== 'number') return false
 *       return true
 *     },
 *   },
 * )
 * ```
 *
 * @example
 * ```typescript
 * // Automatically change a key based on another
 * import { storeObject } from 'ls-proxy'
 *
 * interface Person {
 *   name: string
 *   age: number
 *   minor: boolean
 * }
 *
 * const myPerson = storeObject<Person>(
 *   'myPerson',
 *   {
 *     name: 'Ellie',
 *     age: 17,
 *     minor: true,
 *   },
 *   {
 *     // If the person's age is 18 or greater, set minor to false.
 *     // Otherwise, set it to true.
 *     // This will affect values as they're being stored in localStorage
 *     // and retrieved from it
 *     modify(value) {
 *       if (value.age >= 18) value.minor = false
 *       else value.minor = true
 *       return value
 *     },
 *   },
 * )
 *
 * myPerson.age = 18
 * console.log(myPerson.minor) // false
 * myPerson.age = 16
 * console.log(myPerson.minor) // true
 * ```
 */
function storeObject(lsKey, defaults, configuration = {}) {
    const { checkGets, checkDefaults, mutateProxiedObject, partial, set, get, validate, modify, parse, stringify, } = defaultStoreObjectConfig(configuration);
    /** Call validOrThrow with relevant parameters by default */
    const vot = (value, action = 'set') => validOrThrow(validate, modify, value, action, lsKey);
    const checkParse = (value) => {
        const parsed = parse(value);
        const valid = vot(parsed, 'get');
        return valid;
    };
    const checkStringify = (value) => stringify(vot(value));
    const filterWanted = (obj, defaultIfUndefined = true) => {
        let desiredObject = {};
        Object.keys(defaults).forEach(
        // Set to value found in localStorage if it exists, otherwise use provided default
        key => {
            var _a;
            return (desiredObject[key] = defaultIfUndefined
                ? // Use default if defaultInDefined
                    (_a = obj[key]) !== null && _a !== void 0 ? _a : defaults[key]
                : // Use given value even if undefined
                    obj[key]);
        });
        return desiredObject;
    };
    let object = Object.assign({}, defaults);
    // Update localStorage value or read existing values
    if (checkDefaults) {
        const value = get(lsKey);
        if (value === null) {
            set(lsKey, checkStringify(defaults));
        }
        else if (partial) {
            const current = parse(value);
            object = filterWanted(current);
            const validModified = vot(object);
            set(lsKey, stringify(Object.assign(Object.assign({}, current), validModified)));
        }
        else {
            object = checkParse(value);
        }
    }
    /** Proxy handler for the main object */
    const proxyHandler = {
        set(target, key, value) {
            const setResult = mutateProxiedObject
                ? Reflect.set(target, key, value)
                : true;
            if (partial) {
                const validModified = vot(target);
                set(lsKey, stringify(Object.assign(Object.assign({}, parse(get(lsKey))), validModified)));
            }
            else
                set(lsKey, checkStringify(target));
            return setResult;
        },
        get(target, key) {
            var _a;
            if (checkGets) {
                let newVal = target[key];
                if (partial) {
                    newVal = target[key] = vot(filterWanted(parse(get(lsKey)), false), 'get')[key];
                    vot(target, 'get');
                }
                else {
                    newVal = (_a = checkParse(get(lsKey))[key]) !== null && _a !== void 0 ? _a : defaults[key];
                }
                if (shouldObjectProxy(newVal)) {
                    // Return a Proxy to the object to catch sets
                    return nestedProxyHandler(target, key, newVal, this.set);
                }
                if (mutateProxiedObject) {
                    target[key] = newVal;
                }
                else {
                    return newVal;
                }
            }
            return Reflect.get(target, key);
        },
    };
    return new Proxy(object, proxyHandler);
}
exports.storeObject = storeObject;
const defaultStoreSeparateConfig = ({ id, checkGets, checkDefaults, set, get, validate, modify, parse, stringify, }) => (Object.assign({ id, validate: validate !== null && validate !== void 0 ? validate : (() => true), modify: modify !== null && modify !== void 0 ? modify : (value => value) }, commonDefaults({ checkGets, checkDefaults, set, get, parse, stringify })));
/**
 * Validate and modify a value
 *
 * @param validate Function to validate the object
 * @param modify Function to modify the object
 * @param object The object to modify
 * @param action Whether the object is being get or set
 * @param key The key being validated/modified
 * @template O The stored object
 * @returns The value, if valid
 */
const validOrThrowSeparate = (validate, modify, object, action, key) => {
    const error = new TypeError(action === 'get'
        ? `Validation failed while parsing ${key} from localStorage`
        : `Validation failed while setting to ${key} in localStorage`);
    const valid = validateResult(validate(object, action, key), error);
    if (!valid[0])
        throw valid[1];
    return modify(object, action, key);
};
/**
 * Set multiple individual keys in localStorage with one object
 *
 * @param defaults The defaults values if they are undefined
 * @param configuration Config options
 * @template O The stored object
 *
 * @example
 * ```typescript
 * // No validation
 * import { storeSeparate } from 'ls-proxy'
 *
 * const myObj = storeSeparate({
 *   foo: 'bar',
 *   abc: 123,
 *   numbers: [1, 2, 3],
 * })
 *
 * myObj.foo = 'baz' // Updates localStorage
 * console.log(myObj.foo) // Checks localStorage if checkGets is true
 * console.log(myObj.abc === localStorage.abc) // true
 * ```
 *
 * @example
 * ```typescript
 * // Validating that the key being set/get is correct
 * import { storeSeparate } from 'ls-proxy'
 *
 * const myObj = storeSeparate(
 *   {
 *     foo: 'bar',
 *     abc: 123,
 *   },
 *   {
 *     validate(value, action, key) {
 *       if (key !== 'foo' && key !== 'abc') return false
 *       return true
 *     },
 *   },
 * )
 * ```
 *
 * @example
 * ```typescript
 * // Using IDs to avoid conflicting names
 * import { storeSeparate } from 'ls-proxy'
 *
 * const obj1 = storeSeparate({ foo: 'bar' }, { id: 'obj1' })
 * const obj2 = storeSeparate({ foo: 123 }, { id: 'obj2' })
 *
 * console.log(obj1.foo) // bar
 * console.log(obj2.foo) // 123
 * console.log(localStorage['obj1.foo']) // "bar"
 * console.log(localStorage['obj2.foo']) // 123
 * ```
 *
 * @example
 * ```typescript
 * // Automatically change a key while being set/get
 * import { storeSeparate } from 'ls-proxy'
 *
 * const myObj = storeSeparate(
 *   { base64Value: 'foo' },
 *   {
 *     // Decode base64 on get
 *     get(key) {
 *       return window.btoa(localStorage.getItem(key)!)
 *     },
 *     // Encode base64 on set
 *     set(key, value) {
 *       localStorage.setItem(key, window.atob(value))
 *     },
 *   },
 * )
 *
 * myObj.base64Value = 'bar' // Encoded in localStorage
 * console.log(myObj.base64Value) // Logs 'bar', decoded from localStorage
 * ```
 */
function storeSeparate(defaults, configuration = {}) {
    const { id, checkGets, checkDefaults, mutateProxiedObject, set, get, validate, modify, parse, stringify, } = defaultStoreSeparateConfig(configuration);
    const object = Object.assign({}, defaults);
    /** Call validOrThrow with relevant parameters by default */
    const vot = (key, value, action) => validOrThrowSeparate(validate, modify, { [key]: value }, action, key)[key];
    // Set defaults
    if (checkDefaults) {
        for (const [key, value] of Object.entries(defaults)) {
            const keyPrefix = addId(key, id);
            const lsValue = get(keyPrefix);
            if (lsValue === null) {
                set(keyPrefix, stringify(vot(key, value, 'set')));
            }
            else
                object[key] = vot(parse(lsValue), key, 'get');
        }
    }
    return new Proxy(object, {
        set(target, key, value) {
            // Modify object
            const modified = vot(key, value, 'set');
            set(addId(key, id), stringify(modified));
            if (mutateProxiedObject) {
                return Reflect.set(target, key, modified);
            }
            else {
                return true;
            }
        },
        get(target, key) {
            let newVal = target[key];
            if (checkGets) {
                const valueUnparsed = get(addId(key, id));
                const value = valueUnparsed !== null ? parse(valueUnparsed) : defaults[key];
                newVal = vot(key, value, 'get');
            }
            if (shouldObjectProxy(newVal)) {
                // Return a Proxy to the object to catch sets
                return nestedProxyHandler(target, key, newVal, this.set);
            }
            if (mutateProxiedObject) {
                target[key] = newVal;
            }
            else {
                return newVal;
            }
            return Reflect.get(target, key);
        },
    });
}
exports.storeSeparate = storeSeparate;
const addId = (key, id) => id ? `${id}.${key}` : key;

})();

window.LSProxy = __webpack_exports__;
/******/ })()
;